src/sds.c

```diff
@@ -307,44 +307,13 @@ sds sdsMakeRoomForNonGreedy(sds s, size_t addlen) {
  * After the call, the passed sds string is no longer valid and all the
  * references must be substituted with the new pointer returned by the call. */
 sds sdsRemoveFreeSpace(sds s) {
-    void *sh, *newsh;
-    char type, oldtype = s[-1] & SDS_TYPE_MASK;
-    int hdrlen, oldhdrlen = sdsHdrSize(oldtype);
-    size_t len = sdslen(s);
-    size_t avail = sdsavail(s);
-    sh = (char*)s-oldhdrlen;
-
-    /* Return ASAP if there is no space left. */
-    if (avail == 0) return s;
-
-    /* Check what would be the minimum SDS header that is just good enough to
-     * fit this string. */
-    type = sdsReqType(len);
-    hdrlen = sdsHdrSize(type);
-
-    /* If the type is the same, or at least a large enough type is still
-     * required, we just realloc(), letting the allocator to do the copy
-     * only if really needed. Otherwise if the change is huge, we manually
-     * reallocate the string to use the different header type. */
-    if (oldtype==type || type > SDS_TYPE_8) {
-        newsh = s_realloc(sh, oldhdrlen+len+1);
-        if (newsh == NULL) return NULL;
-        s = (char*)newsh+oldhdrlen;
-    } else {
-        newsh = s_malloc(hdrlen+len+1);
-        if (newsh == NULL) return NULL;
-        memcpy((char*)newsh+hdrlen, s, len+1);
-        s_free(sh);
-        s = (char*)newsh+hdrlen;
-        s[-1] = type;
-        sdssetlen(s, len);
-    }
-    sdssetalloc(s, len);
-    return s;
+    return sdsResize(s, sdslen(s));
 }
 
 /* Resize the allocation, this can make the allocation bigger or smaller,
- * if the size is smaller than currently used len, the data will be truncated */
+ * if the size is smaller than currently used len, the data will be truncated.
+ * To avoid repeated calls, the sds allocation size will be set to the requested size
+ * regardless of the actual allocation size. */
 sds sdsResize(sds s, size_t size) {
```

Comment:
GH won't let me comment on this:
```c
    /* Don't use type 5, it is not good for strings that are resized. */
    if (type == SDS_TYPE_5) type = SDS_TYPE_8;
```
i'm having a second thought about unifying these functions due to this part.
sdsResize is meant to be used on an sds that constantly changes size, and thus we want to avoid TYPE_5.
but sdsRemoveFreeSpace, specifically in our case, is used on a string that's gonna become immutable from now on.
and keeping the extra 2 bytes on very short strings can be an overhead this PR adds on memory usage.

Reply:
After this change, we call `sdsRemoveFreeSpace` from `tryObjectEncoding` only for >32KB strings. 
Additionally, it is called for the querybuf, which is not immutable.
Are you referring to the function being used elsewhere, such as in `RM_TrimStringAllocation`?
If so, should I revert this change or wrap the function with an argument indicating whether the string is immutable?

Reply:
it's about all calls to trimStringObjectIfNeeded and sdsRemoveFreeSpace.

first, i now notice the change we did in tryObjectEncoding probably needs to be reflected to RedisModuleCommandDispatcher (or better yet, moved to inside trimStringObjectIfNeeded so it serves both).

and secondly, our change in sdsRemoveFreeSpace (to avoid using TYPE_5), is affecting RM_StringTruncate.
i think adding an explicit argument would be better, since otherwise (if we revert) people can have a hard time understanding the differences between these two functions.
the only thing that bothers me is that if these functions are heavily used, i don't want to add another argument in a ton of places. luckily neither of them is heavily used.

WDYT? am i missing something or looking too much into it?

Reply:
I think adding an argument for it makes sense, we use both `sdsResize` and  `sdsRemoveFreeSpace`  only in 4 places, and `sdsRemoveFreeSpace` will still have only 2 arguments after the change.
Changed accordingly.